streams/src/main/java/org/apache/kafka/streams/processor/internals/TaskManager.java

```diff
@@ -233,26 +218,51 @@ InternalTopologyBuilder builder() {
     }
 
     /**
-     * Similar to shutdownTasksAndState, however does not close the task managers, in the hope that
-     * soon the tasks will be assigned again
+     * Closes suspended active tasks and standby tasks that were reassigned at the end of a rebalance.
+     *
+     * @return list of changelog topic partitions from revoked tasks
      * @throws TaskMigratedException if the task producer got fenced (EOS only)
      */
-    void suspendTasksAndState()  {
-        log.debug("Suspending all active tasks {} and standby tasks {}", active.runningTaskIds(), standby.runningTaskIds());
+    List<TopicPartition> closeRevokedTasks() {
+        log.debug("Closing revoked active tasks {} and standby tasks {}", revokedActiveTaskIds(), revokedStandbyTaskIds());
+
+        final List<TopicPartition> revokedChangelogPartitions = new ArrayList<>();
 
+        active.closeNonAssignedSuspendedTasks(assignedActiveTasks, revokedChangelogPartitions);
+        standby.closeReassignedStandbyTasks(revokedStandbyTasks, revokedChangelogPartitions);
+
+        // remove any revoked partitions from the restore consumer's assignment and changelog reader
+        final Set<TopicPartition> updatedAssignment = new HashSet<>(restoreConsumer.assignment());
```

Comment:
Note that the restore-consumer is shared for both restoring active tasks, as well as for updating standby tasks, and it would only start updating standby tasks until all active tasks has completed restoring and now in running state.

This means at a given time, the restore-consumer would only be assigned with restoring active tasks, or with standby tasks; on the other hand, the `changelogReader` is only used for restoring active tasks.

So far this logic is still correct, but I'd suggest we handle such two cases like `closeSuspendedActiveTasks` and `closeStandbyTasks` separately, and also clear standby records / reassign restore-consumer separately.

Reply:
More specifically, during `onPartitionsRevoked` the restore-consumer is assigned either:

1) with active restoring task partitions, or
2) with standby task partitions

With 1), upon `onPartitionsAssigned` we would continue restoring active tasks while no standby tasks would be processed yet, hence we could update restore-consumer with revoked partitions and newly added task partitions that needs restoring; 

With 2), upon `onPartitionsAssigned` there are two cases: 1) if there are newly added task partitions that needs restoring, we should "revoke" all standby tasks since we need to first go to restore those active tasks; 2) no new active tasks need restoring, we should only "revoke" those standby tasks that are re-assigned.

Reply:
The shared restore consumer is kind of a pain, but here are my thoughts:
1) AFAICT restored partitions are not removed from the assignment (or even paused) until they are all completed and replaced with the standby changelogs. We *need* to remove closed/reassigned partitions, and *should* remove/pause completely restored ones.
2) After a rebalance, we should clear any standby partitions from the restore consumer if we have any new tasks that might need restoring. The restoreConsumer assigns new partitions incrementally (see `StoreChangelogReader#startRestoration`) so we must be careful not to mix these with standbys.

Just pushed these changes, also split up closing the suspended and standbys. I tried to follow the restore consumer code around and believe it will be properly assigned/updated but lmk if you see any potential issues.